torchtune/modules/kv_cache.py

```diff
@@ -17,50 +17,38 @@ class KVCache(nn.Module):
     Args:
         max_batch_size (int): maximum batch size model will be run with
         max_seq_len (int): maximum sequence length model will be run with
-        n_kv_heads (int): number of kv heads
+        num_heads (int): number of heads. We take num_heads instead of num_kv_heads because
+            the cache is created after we've expanded the key and value tensors to have the
+            same shape as the query tensor. See attention.py for more details
         head_dim (int): per-attention head embedding dimension
+        dtype (torch.dtype): dtype for the caches
     """
 
     def __init__(
         self,
         max_batch_size: int,
         max_seq_len: int,
-        n_kv_heads: int,
+        num_heads: int,
         head_dim: int,
-    ):
+        dtype: torch.dtype,
+    ) -> None:
         super().__init__()
-        cache_shape = (max_batch_size, max_seq_len, n_kv_heads, head_dim)
-        self.register_buffer("k_cache", torch.zeros(cache_shape), persistent=False)
-        self.register_buffer("v_cache", torch.zeros(cache_shape), persistent=False)
+        cache_shape = (max_batch_size, num_heads, max_seq_len, head_dim)
+        self.register_buffer(
+            "k_cache", torch.zeros(cache_shape, dtype=dtype), persistent=False
+        )
+        self.register_buffer(
+            "v_cache", torch.zeros(cache_shape, dtype=dtype), persistent=False
+        )
         self.max_batch_size = max_batch_size
 
-    def update(
-        self, bsz: int, seq_len: int, curr_pos: int, k_val: Tensor, v_val: Tensor
-    ) -> Tuple[Tensor, Tensor]:
-        """
-        Updates the kv-cache at curr_pos with the given k_val and v_val.
-
-        Args:
-            bsz (int): Batch size.
-            seq_len (int): Sequence length.
-            curr_pos (int): Current position in sequence.
-            k_val (Tensor): New k value.
-            v_val (Tensor): New v value.
+    def update(self, input_pos, k_val, v_val) -> Tuple[Tensor, Tensor]:
+        # input_pos: [S], k_val: [B, H, S, D]
+        assert input_pos.shape[0] == k_val.shape[2]
 
-        Raises:
-            ValueError: if bsz is greater than the ``max_batch_size`` supported by the model
+        k_out = self.k_cache
+        v_out = self.v_cache
+        k_out[:, :, input_pos] = k_val
+        v_out[:, :, input_pos] = v_val
```

Comment:
If I understand correctly, is it now the case that `max_batch_size` really just means batch size? If so, maybe we should update its name

Reply:
I don't know why it was called max_batch_size ever, but this seems to be the convention. Let me do some more research on this